# **Workshop: Capturing Human Perception of Neighborhoods Using Online Data with Language and Vision Models**

---

## **Data Exploration: Understanding the StreetLens Dataset**

In this session, we will explore the StreetLens dataset.

This aims to highlight how each part of the dataset supports StreetLens’s researcher-oriented workflow.

We will walk through the following resources:

- **[1] Codebook**
  - Contain definitions and criteria for scoring  
- **[2] Human Annotation Scores**
  - Contain labeled evaluations generated by human coders
- **[3] Related Research Papers**
  - Provide domain knowledge and theoretical context
- **[4] Street View Image**
  - Provide visual assessments using images retrieved from the Google Street View Image API


>  **Human annotation scores** were generated with a trained coder virtually assessing street segments using Google Earth Pro. This method is introduced and validated by Pasco and White. </n>

> **Note**: The datasets used in this workshop are for demonstration purposes only and may not be shared outside this environment.


## Before we start:
Make sure to save a copy of this Jupyter notebook to your own Google Drive,
so that you can make edits and run the code yourself.

Go to a tab **File → Save a copy in Drive**, then run the code in your copy.

<img src = "https://drive.google.com/uc?id=1RtVrYCxlT-nsNNrBWvuzfwfHgZkaXnzV" height = 400 width = 650>

# Environment Setup for Data Exploration

In [1]:
from google.colab import drive
drive.mount('/content/Shareddrives')

Mounted at /content/Shareddrives


In [2]:
%cd /content/Shareddrives/Shareddrives/UCGIS2025-Workshop-CSENG-Computer\ Science
!pip install pillow ipyplot numpy
import json
import os
import pandas as pd
from tabulate import tabulate
from streetlens.data_sample_loader import *

/content/Shareddrives/Shareddrives/UCGIS2025-Workshop-CSENG-Computer Science
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 48.6 MB/s eta 0:00:00

        WARNING! Google Colab Environment detected!
        You might encounter issues while running in Google Colab environment.
        If images are not displaying properly please try setting `force_b64` param to `True`.
        


# [1] Codebook
In this section, we will examine the codebook, which outlines the criteria used to assess neighborhood environments.

The original PDF format has been converted into a structured JSON format to integrate into our workflow.

<img src = "https://drive.google.com/uc?id=1_ZpuC4cgHq8gzOv4Oaj1iUnpR9tlTi5S" height = 400 width = 650>

In [3]:
codebook_path = './dataset/annotation/sso_codebook.json'


DataSampleloader.load_codebook(codebook_path)

{'Disorder 3': {'question': 'Is there graffiti on buildings, signs, walls, fences, sidewalks, poles, or streets (does not include spray painted measurements shown below)?', 'answer_options': '1 - Yes 0 - No'}, 'Decay 1': {'question': 'Condition of the visible Street', 'answer_options': '1- Good (majority of the street has minimal slight cracks, either no potholes or minimal potholes are fixed/covered) 2 - Fair (some moderate cracks, a mix of open and recently filled potholes and cracks) 3 - Poor (Sizable cracks, potholes, or broken curbs along MOST or ALL of the street; cracks and potholes have not been recently filled in) Under construction Street is present, but cannot evaluate'}, 'Decay 2': {'question': 'Condition of the sidewalk', 'answer_options': '1 - Good (NO holes, sizable cracks, or crumbling or uneven pavement) 2 - Fair (Holes, sizable cracks, or crumbling or uneven pavement, outgrown weeds along SOME of the side walk) 3 - Poor (Holes, sizable cracks, or crumbling or uneven p

# [2] Human Annotation Scores

In this section, we will explore human annotation scores.

A human coder evaluates each street block and its specific direction using street view imagery, guided by criteria defined in the codebook.


In [4]:
annotation_path = './dataset/annotation/sso_annotation.csv'

DataSampleloader.load_annotation(annotation_path)

+----+-------------------+----------------------------------+--------------------------+------------+-------+-------+-------+-------+-------+-------+-------+-----------+--------------+--------------+--------------+-----------+
|    | Street Block ID   | Direction of Target Block Face   | Picture Date (mm/year)   |    Decay 1 |   SS1 |   SS2 |   SS3 |   SS4 |   SS5 |   SS6 |   SS7 |   Decay 2 |   Disorder 1 |   Disorder 2 |   Disorder 3 |   Decay 3 |
|----+-------------------+----------------------------------+--------------------------+------------+-------+-------+-------+-------+-------+-------+-------+-----------+--------------+--------------+--------------+-----------|
|  0 | 62146             | W                                | Jul-21                   |          3 |     1 |     1 |     1 |     1 |     1 |     1 |     1 |         2 |            2 |            0 |            1 |       nan |
|  1 | 281, 282          | W                                | Aug-21                   |    

# [3] Related Research Papers

In this section, we will explore research papers that offer foundational domain knowledge for our workflow.

This enables to guide our system in adopting an appropriate role for domain-specific tasks. To incorporate these knowledges into our workflow, we’ve converted the paper abstracts into a JSON format.

<img src = "https://drive.google.com/uc?id=1OlEAuqZ-a6DcdoQQX641jwE2pD-KVEBe" height = 400 width = 600>




In [5]:
paper_abs_path = './dataset/paper/abstract.json'

DataSampleloader.loab_paper(paper_abs_path)

+----+-----------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

#[4] Street View Image

In this section, we will examine street view image.

The following steps outline how we explore street view image:
- [4-1] Access pre-downloaded street view images
- [4-2] View street view images for each street block

### [4-1] Access pre-downloaded street view images

We have pre-downloaded street view images for a few street block. Each folder is named after the block ID (a unique ID representing the street block), which you can find details in the table below.

- Clicking on the **block ID** will take you to the Google Drive folder containing the images
- Clicking on the **approximate center point** will open the location in Google Maps.

| Block ID | Street Address | Approximate Center Point |
| --- | --- | --- |
| [281](https://drive.google.com/drive/folders/1fAm659Hzz299TtHXIwxHBb9kcaKfXYSd?usp=drive_link) | N Humboldt Ave, MN | [(45.077581, -93.299938)](https://maps.app.goo.gl/AoDfoT8hRJn9qgR9A) |
| [282](https://drive.google.com/drive/folders/1EfyIfn5hKzSGv58XUxotQ-9a_2AZ3Y9U?usp=drive_link) | N Humboldt Ave, MN | [(45.078889, -93.300085)](https://maps.app.goo.gl/CeMCoJTnXchbnAQk6) |
| [9576](https://drive.google.com/drive/folders/1DBv0ZmHSUKbLr0F9V4S6ZAom3Uas3S-r?usp=drive_link) | Humboldt Pl N, MN | [(45.077768, -93.300613)](https://maps.app.goo.gl/6e9YLomgC6yVv1PU9) |
| [62146](https://drive.google.com/drive/folders/138Z8uyQeyKJdmW7Vx59wgNJNYwexbSlR?usp=drive_link) | Minnehaha Ave, MN | [(44.954241, -93.238921)](https://maps.app.goo.gl/jATpc2fqMXbMbzyL8) |



### [4-2] View street view images for each street block
To view street view images from any of the blocks indicated on the table above, replace the value after the equal sign in the following code with the **block ID**. The next code block will show all available images for that street block.

In [6]:
import os
import re
import ipyplot    # Utility to display images
from PIL import Image
import numpy as np

img_dir = "/content/Shareddrives/Shareddrives/UCGIS2025-Workshop-CSENG-Computer Science/dataset/img"
blockID = 281

DataSampleloader.load_svi(img_dir,blockID)

Output hidden; open in https://colab.research.google.com to view.

# [Optional] Query Street View Image from Google Street View API

We also provide a script for querying street view imagery from the Google Street View API, if you are curious about how to play around with Street View imagery!


Install packages  

In [ ]:
!pip install google_streetview pillow

  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Created wheel for google_streetview: filename=google_streetview-1.2.9-py3-none-any.whl size=9777 sha256=a8a929ee54eb01fc3e487e827c9b7b8223fb772bb7a24e54406fdcf288017799
  Stored in directory: /root/.cache/pip/wheels/8f/55/d0/074e47d0e3fede14e60ddcd7b1a59681e1f1c3fd5b56cef79d
  Created wheel for kwconfig: filename=kwconfig-1.1.7-py3-none-any.whl size=4974 sha256=245f0702aa85e242edfb77d037cf61223aa7c922c7d5b00b2adc15c50cf518bc
  Stored in directory: /root/.cache/pip/wheels/5d/ae/f3/4f084ead544ae0187acf5ef586c5ee24ec92e1029b3383a8ac
Successfully built google_streetview kwconfig


Then, visit the [Google SVI API webpage](https://developers.google.com/maps/documentation/streetview/overview) and follow the steps to create a key.

Please replace the value after the equal sign in the following code with your key.

In [ ]:
streetview_api = "YOUR_API_KEY"

Set the **latitude and longitude** for the location you want, and specify the **direction** (a.k.a. "heading") using the following angle system:


*   North = 0
*   East = 90
*   South = 180
*   West = 270

In [ ]:
# Example latitude/longitude and direction

latitude = 44.965581352431904
longitude = -93.20771509754005
angle_of_view = 90

Download street view image from Google Street View API

In [ ]:
import time
import json
import google_streetview.api

param = [{'size': '640x640',                                   # default to max pixel size
         'location': f'{float(latitude)},{float(longitude)}',
         'pitch': '0',                                         # vertical view (just 0)
         'heading': f'{angle_of_view}',                        # angle control
         'key': streetview_api
         }]

results = google_streetview.api.results(param)

save_dir = "YOUR_OUTPUT_DIR"
results.download_links(dir_path=save_dir, metadata_file=f"metadata.json")

NameError: name 'streetview_api' is not defined

Details about the downloaded street view image

In [ ]:
import json
from PIL import Image

with open(f"{save_dir}/metadata.json") as f:
    d = json.load(f)

print("Image details:")

for key, value in d[0].items():
  print(f"{key}: {value}")

print("Image:")
im = Image.open(f"{save_dir}/*.jpg")